lib/runcmd.c

```diff
@@ -115,27 +116,133 @@ pid_t runcmd_pid(int fd)
 #define STATE_INDQ  (1 << 3)
 #define STATE_SPECIAL (1 << 4)
 #define STATE_BSLASH (1 << 5)
+#define STATE_INVAR (1 << 6)
+#define STATE_INVAL (1 << 7)
 #define in_quotes (state & (STATE_INSQ | STATE_INDQ))
 #define is_state(s) (state == s)
 #define set_state(s) (state = s)
 #define have_state(s) ((state & s) == s)
 #define add_state(s) (state |= s)
 #define del_state(s) (state &= ~s)
 #define add_ret(r) (ret |= r)
-int runcmd_cmd2strv(const char *str, int *out_argc, char **out_argv)
+int runcmd_cmd2strv(const char *str, int *out_argc, char **out_argv, int *out_envc, char **out_env)
 {
-	int arg = 0, a = 0;
+	int arg = 0, a = 0, env = 0, e = 0, argstart = 0;
 	unsigned int i;
 	int state, ret = 0;
 	int seen_space = 0;
 	size_t len;
-	char *argz;
+	char *argz, *envz;
 
+	// extract leading environment variables
 	set_state(STATE_NONE);
 	len = strlen(str);
+	envz = malloc(len + 1);
+	for (i = 0; i < len; i++) {
+		const char *p = &str[i];
+
+		switch (*p) {
+		case 0:
+			return ret;
+
+		case ' ': case '\t': case '\r': case '\n':
+			if(in_quotes)
+				break;
+			if(is_state(STATE_INVAR)) {
+				/* abort checking for variables, this is something else */
+				env--;
+				i = len;
```

Comment:
Is this different from just doing a break? We set `i = len`, then go on to the next loop iteration, with the loop condition `i < len` guaranteed to fail?

Reply:
yes, `break` breaks from the current switch and still finishes the for loop (which appends the current character at the end of the loop). But i could add a simple flag, maybe this would make the code more clear.

Reply:
Basically, we use "break" to break the switch and add a character and "continue" to do nothing and start over with the next character.